app/code/Magento/AwsS3/Model/Cached/CachedAdapter.php

```diff
@@ -0,0 +1,308 @@
+            return $result;
+        }
+
+        $result = $this->adapter->read($path);
+        $this->cache->updateObject($path, ['contents' => $result], true);
+
+        return $result;
+    }
+
+    /**
+     * @inheritdoc
+     */
+    public function readStream(string $path)
+    {
+        return $this->adapter->readStream($path);
+    }
+
+    /**
+     * @inheritdoc
+     */
+    public function listContents(string $path, bool $deep): iterable
+    {
+        if ($this->cache->isComplete($path, $deep)) {
+            return $this->cache->listContents($path, $deep)->getIterator();
+        }
+
+        $contents = $this->adapter->listContents($path, $deep);
+        $objects = [];
+        while ($contents->valid()) {
+            $objects[] = $contents->current();
+            $contents->next();
+        }
+        if ($objects) {
+            $this->cache->storeContents($path, $objects, $deep);
+        }
+
+        return $objects;
+    }
+
+    /**
+     * @inheirtdoc
+     */
+    public function lastModified(string $path): FileAttributes
+    {
+        $result = $this->cache->lastModified($path);
+        if ($result) {
+            return new FileAttributes($path, null, null, $result);
+        }
+
+        $result = $this->adapter->lastModified($path);
+        $object = $result->jsonSerialize() + compact('path');
+        $this->cache->updateObject($path, $object, true);
+
+        return $result;
+    }
+
+    /**
+     * @inheirtdoc
+     */
+    public function fileSize(string $path): FileAttributes
+    {
+        $result = $this->cache->fileSize($path);
+        if ($result) {
+            return new FileAttributes($path, $result);
+        }
+
+        $result = $this->adapter->fileSize($path);
+        $object = $result->jsonSerialize() + compact('path');
+        $this->cache->updateObject($path, $object, true);
+
+        return $result;
+    }
+
+    /**
+     * @inheirtdoc
+     */
+    public function mimeType(string $path): FileAttributes
+    {
+        $result = $this->cache->mimeType($path);
+        if ($result) {
+            return new FileAttributes($path, null, null, null, $result);
+        }
+
+        $result = $this->adapter->mimeType($path);
+        $object = $result->jsonSerialize() + compact('path');
+        $this->cache->updateObject($path, $object, true);
+
+        return $result;
+    }
+
+    /**
+     * @inheirtdoc
+     */
+    public function visibility(string $path): FileAttributes
+    {
+        $result = $this->cache->visibility($path);
+        if ($result) {
+            return new FileAttributes($path, null, $result);
+        }
+
+        $result = $this->adapter->visibility($path);
+        $object = $result->jsonSerialize() + compact('path');
+        $this->cache->updateObject($path, $object, true);
+
+        return $result;
+    }
+
+    /**
+     * Get file metadata.
+     *
+     * @deplacated There is no getMetadata() method in FilesystemAdapter anymore.
```

Comment:
Do we need to introduce this deplacated method?

Reply:
There are several clients for this method:
Magento\Eav\Model\Attribute\Data\Image, Magento\MediaGallerySynchronization\Model\CreateAssetFromFile, Magento\Framework\File\Mime.

Reply:
The comment and `@deplacated` annotation should be removed.

Is implementing `League\Flysystem\FilesystemAdapter` necessary?